File: source/python/cycle_detection.py
def has_cycle(adj, n):
    color = [0] * n  # 0=white, 1=gray, 2=black

    for start in range(n):
        if color[start] != 0:
            continue
        stack = [{"node": start, "idx": 0}]
        color[start] = 1

        while stack:
            top = stack[-1]
            u = top["node"]
            neighbors = adj[u]

            if top["idx"] < len(neighbors):
                v = neighbors[top["idx"]]
                top["idx"] += 1
                if color[v] == 1:
                    return True
                if color[v] == 0:
                    color[v] = 1
                    stack.append({"node": v, "idx": 0})
            else:
                color[u] = 2
                stack.pop()

    return False

adj1 = [[1], [2], [0, 3], []]
if has_cycle(adj1, 4):
    print("graph1: cycle detected")
else:
    print("graph1: no cycle")

adj2 = [[1, 2], [3], [3], []]
if has_cycle(adj2, 4):
    print("graph2: cycle detected")
else:
    print("graph2: no cycle")

File: source/rust/cycle_detection.rs
fn has_cycle(adj: &Vec<Vec<usize>>, n: usize) -> bool {
    let mut color = vec![0u8; n]; // 0=white, 1=gray, 2=black

    for start in 0..n {
        if color[start] != 0 { continue; }
        let mut stack: Vec<(usize, usize)> = vec![(start, 0)];
        color[start] = 1;

        while let Some(top) = stack.last_mut() {
            let u = top.0;
            let idx = top.1;
            let neighbors = &adj[u];

            if idx < neighbors.len() {
                let v = neighbors[idx];
                stack.last_mut().unwrap().1 += 1;
                if color[v] == 1 { return true; }
                if color[v] == 0 {
                    color[v] = 1;
                    stack.push((v, 0));
                }
            } else {
                color[u] = 2;
                stack.pop();
            }
        }
    }
    false
}

fn main() {
    let adj1 = vec![vec![1], vec![2], vec![0, 3], vec![]];
    if has_cycle(&adj1, 4) { println!("graph1: cycle detected"); } else { println!("graph1: no cycle"); }

    let adj2 = vec![vec![1, 2], vec![3], vec![3], vec![]];
    if has_cycle(&adj2, 4) { println!("graph2: cycle detected"); } else { println!("graph2: no cycle"); }
}
